Ginger/GingerCoreCommon/Repository/BusinessFlowLib/ActivitiesGroup.cs

```diff
@@ -204,49 +204,143 @@ public override string ItemName
-                        int originalIndex = ((BusinessFlow)hostItem).ActivitiesGroups.IndexOf(activitiesGroupInstance);
-                        ((BusinessFlow)hostItem).ActivitiesGroups.Remove(activitiesGroupInstance);
-                        ((BusinessFlow)hostItem).SetUniqueActivitiesGroupName(newInstance);
-                        ((BusinessFlow)hostItem).ActivitiesGroups.Insert(originalIndex, newInstance);
-                        ((BusinessFlow)hostItem).AttachActivitiesGroupsAndActivities();
+                        Activity updatedAct = null;
+
+                        // Activity still exist in the group, thus re-add the same activity to the group
+                        updatedAct = existingActivities.Where(a => a.Guid == actID.ActivityGuid).FirstOrDefault();
+
+                        // In case, group was Replaced/Overwritten
+                        if (updatedAct == null)
+                        {
+                            updatedAct = existingActivities.Where(a => a.ParentGuid == actID.ActivityGuid).FirstOrDefault();
+                        }
+
+                        if (updatedAct != null)
+                        {
+                            currentBF.AddActivity(updatedAct, activitiesGroupInstance, insertIndex);         //  .Activities.Insert(firstActivityIndex, updatedAct);
+                            insertIndex++;
+                            existingActivities.Remove(updatedAct);
+                        }
+                        // Activity doesn't exist in the group and the shared group is recently updated by addition of this activity, thus add this activity to the group instance in the Business Flow
+                        else if(extraDetails != null)
+                        {
+                            updatedAct = (extraDetails as ObservableList<Activity>).Where(a => a.ActivityName == actID.ActivityName && a.Guid == actID.ActivityGuid).FirstOrDefault();
+                            if (updatedAct == null)
+                            {
+                                updatedAct = (extraDetails as ObservableList<Activity>).Where(a => a.Guid == actID.ActivityGuid).FirstOrDefault();
+                            }
+                            if(updatedAct == null)
+                            {
+                                updatedAct = (extraDetails as ObservableList<Activity>).Where(a => a.ActivityName == actID.ActivityName).FirstOrDefault();
+                            }
+
+                            if (updatedAct != null)
+                            {
+                                updatedAct = updatedAct.CreateInstance(true) as Activity;
+                                currentBF.AddActivity(updatedAct, activitiesGroupInstance, insertIndex);
+                                insertIndex++;
+                            }
+                            //currentBF.ImportActivitiesGroupActivitiesFromRepository(activitiesGroupInstance, extraDetails as ObservableList<Activity>);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this commented out code.](https://app.codacy.com/app/Ginger-Automation/Ginger/pullRequest?prid=4582993)